File: packages/http-client-csharp-generator/src/AutoRest.CSharp/Common/Input/InputTypes/InputParameter.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using AutoRest.CSharp.Input;
using AutoRest.CSharp.Output.Builders;
using AutoRest.CSharp.Output.Models.Serialization;

namespace AutoRest.CSharp.Common.Input;

internal record InputParameter(
    string Name,
    string NameInRequest,
    string? Description,
    InputType Type,
    RequestLocation Location,
    InputConstant? DefaultValue,
    VirtualParameter? VirtualParameter,
    InputParameter? GroupedBy,
    InputOperationParameterKind Kind,
    bool IsRequired,
    bool IsApiVersion,
    bool IsResourceParameter,
    bool IsContentType,
    bool IsEndpoint,
    bool SkipUrlEncoding,
    bool Explode,
    string? ArraySerializationDelimiter,
    string? HeaderCollectionPrefix)
{
    public InputParameter() : this(
        Name: string.Empty,
        NameInRequest: string.Empty,
        Description: null,
        Type: InputPrimitiveType.Object,
        Location: RequestLocation.None,
        DefaultValue: null,
        VirtualParameter: null,
        GroupedBy: null,
        Kind: InputOperationParameterKind.Method,
        IsRequired: false,
        IsApiVersion: false,
        IsResourceParameter: false,
        IsContentType: false,
        IsEndpoint: false,
        SkipUrlEncoding: false,
        Explode: false,
        ArraySerializationDelimiter: null,
        HeaderCollectionPrefix: null)
    { }

    private SerializationFormat? _serializationFormat;
    public SerializationFormat SerializationFormat => _serializationFormat ??= GetSerializationFormat(Type, Location);

    private static SerializationFormat GetSerializationFormat(InputType parameterType, RequestLocation requestLocation)
    {
        var affectType = parameterType switch
        {
            InputListType listType => listType.ElementType,
            InputDictionaryType dictionaryType => dictionaryType.ValueType,
            _ => parameterType
        };
        if (affectType is InputPrimitiveType { Kind: InputTypeKind.DateTime })
        {
            if (requestLocation == RequestLocation.Header)
            {
                return SerializationFormat.DateTime_RFC7231;
            }

            if (requestLocation == RequestLocation.Body)
            {
                return SerializationFormat.DateTime_RFC3339;
            }
        }

        return SerializationBuilder.GetSerializationFormat(affectType);
    }
}

File: packages/http-client-csharp-generator/src/TypeSpec.Extension/Emitter.Csharp/src/type/inputParameter.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { InputConstant } from "./inputConstant.js";
import { InputOperationParameterKind } from "./inputOperationParameterKind.js";
import { InputType } from "./inputType.js";
import { RequestLocation } from "./requestLocation.js";

//TODO: Define VirtualParameter for HLC
export interface VirtualParameter {}
export interface InputParameter {
    Name: string;
    NameInRequest: string;
    Description?: string;
    Type: InputType;
    Location: RequestLocation;
    DefaultValue?: InputConstant;
    VirtualParameter?: VirtualParameter; //for HLC, set null for typespec
    GroupedBy?: InputParameter;
    Kind: InputOperationParameterKind;
    IsRequired: boolean;
    IsApiVersion: boolean;
    IsResourceParameter: boolean;
    IsContentType: boolean;
    IsEndpoint: boolean;
    SkipUrlEncoding: boolean;
    Explode: boolean;
    ArraySerializationDelimiter?: string;
    HeaderCollectionPrefix?: string;
}
